### ISSUE TITLE
2.13.3: sphinx warnings

### ISSUE THREAD

**Issue by kloczek** [agronholm/typeguard#250]
**Describe the bug**
First of all  looks like at the moment is not possible to generate documentation using docs/Makefie
```console
[tkloczko@devel-g2v typeguard-2.13.3]$ make -C docs man
make: Entering directory '/home/tkloczko/rpmbuild/BUILD/typeguard-2.13.3/docs'
Running Sphinx v4.5.0
/usr/lib/python3.8/site-packages/pkg_resources/__init__.py:123: PkgResourcesDeprecationWarning: 1.17.1-unknown is an invalid version and will not be supported in a future release
  warnings.warn(

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/sphinx/config.py", line 332, in eval_config_file
    exec(code, namespace)
  File "/home/tkloczko/rpmbuild/BUILD/typeguard-2.13.3/docs/conf.py", line 17, in <module>
    v = pkg_resources.get_distribution('typeguard').parsed_version
  File "/usr/lib/python3.8/site-packages/pkg_resources/__init__.py", line 478, in get_distribution
    dist = get_provider(dist)
  File "/usr/lib/python3.8/site-packages/pkg_resources/__init__.py", line 354, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/usr/lib/python3.8/site-packages/pkg_resources/__init__.py", line 909, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python3.8/site-packages/pkg_resources/__init__.py", line 795, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'typeguard' distribution was not found and is required by the application
```
This can be fixed by patch
```patch
--- a/docs/conf.py~     2021-12-10 21:05:38.000000000 +0000
+++ b/docs/conf.py      2022-04-05 23:54:53.571760525 +0000
@@ -1,6 +1,10 @@
 #!/usr/bin/env python3
 import pkg_resources

+import sys
+import os
+sys.path.append(os.path.abspath('../src'))
+
 extensions = [
     'sphinx.ext.autodoc',
     'sphinx.ext.intersphinx',
```
However after fixeng that sphinx shows few warnings
```console
+ /usr/bin/sphinx-build -n -T -b man docs build/sphinx/man
Running Sphinx v4.5.0
making output directory... done
WARNING: html_static_path entry '_static' does not exist
loading intersphinx inventory from http://docs.python.org/3/objects.inv...
intersphinx inventory has moved: http://docs.python.org/3/objects.inv -> https://docs.python.org/3/objects.inv
building [mo]: targets for 0 po files that are out of date
building [man]: all manpages
updating environment: [new config] 4 added, 0 changed, 0 removed
reading sources... [100%] versionhistory
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
writing... python-typeguard.3 { userguide api versionhistory } /home/tkloczko/rpmbuild/BUILD/typeguard-2.13.3/src/typeguard/__init__.py:docstring of typeguard.typechecked:: WARNING: py:class reference target not found: typeguard.T_CallableOrType
/home/tkloczko/rpmbuild/BUILD/typeguard-2.13.3/src/typeguard/__init__.py:docstring of typeguard.typechecked:: WARNING: py:class reference target not found: typeguard.T_CallableOrType
/home/tkloczko/rpmbuild/BUILD/typeguard-2.13.3/src/typeguard/__init__.py:docstring of typeguard.typechecked:: WARNING: py:class reference target not found: typeguard.T_CallableOrType
/home/tkloczko/rpmbuild/BUILD/typeguard-2.13.3/src/typeguard/__init__.py:docstring of typeguard.typechecked:: WARNING: py:class reference target not found: typeguard.T_CallableOrType
/home/tkloczko/rpmbuild/BUILD/typeguard-2.13.3/src/typeguard/importhook.py:docstring of typeguard.importhook.install_import_hook:5: WARNING: py:class reference target not found: typeguard.importhook.ImportHookManager
done
build succeeded, 6 warnings.
```

**To Reproduce**
Just run `sphinx-build -n -T -b man docs build/sphinx/man`

**Expected behavior**
It should be no warning on generate docuemntation.

**Additional context**
List of modules installed in build env:
```console
Package                       Version
----------------------------- -----------------
alabaster                     0.7.12
attrs                         21.4.0
Babel                         2.9.1
beautifulsoup4                4.10.0
build                         0.7.0
charset-normalizer            2.0.12
codespell                     2.1.0
coverage                      6.3.2
cycler                        0.11.0
distro                        1.7.0
docutils                      0.17.1
extras                        1.0.0
fixtures                      3.0.0
fonttools                     4.32.0
gpg                           1.17.1-unknown
html2text                     2020.1.16
idna                          3.3
imagesize                     1.3.0
importlib-metadata            4.11.3
iniconfig                     1.1.1
Jinja2                        3.1.1
kiwisolver                    1.3.2
libcomps                      0.1.18
MarkupSafe                    2.1.1
matplotlib                    3.5.1
metaextract                   1.0.8
mypy                          0.940
mypy-extensions               0.4.3
numpy                         1.22.3
olefile                       0.46
packaging                     21.3
pbr                           5.8.1
pep517                        0.12.0
Pillow                        9.1.0
pip                           22.0.4
pluggy                        1.0.0
py                            1.11.0
py2pack                       0.8.7
Pygments                      2.11.2
PyGObject                     3.42.0
pyparsing                     3.0.8
pypi-search                   1.2.1
pytest                        7.1.1
pytest-cov                    3.0.0
python-dateutil               2.8.2
pytz                          2022.1
requests                      2.27.1
rpm                           4.17.0
setuptools                    62.0.0
setuptools-scm                6.4.2
six                           1.16.0
snowballstemmer               2.2.0
soupsieve                     2.3.2.post1
Sphinx                        4.5.0.dev20220404
sphinx-autodoc-typehints      1.12.0
sphinxcontrib-applehelp       1.0.2.dev20220410
sphinxcontrib-devhelp         1.0.2.dev20220410
sphinxcontrib-htmlhelp        2.0.0
sphinxcontrib-jsmath          1.0.1.dev20220410
sphinxcontrib-qthelp          1.0.3.dev20220410
sphinxcontrib-serializinghtml 1.1.5
testtools                     2.5.0
tomli                         2.0.1
typing_extensions             4.1.1
urllib3                       1.26.9
wheel                         0.37.1
zipp                          3.8.0
```

**Comment by agronholm** [comment]
The project needs to be installed for this to work (even a "develop" install will do). Have you done that? That patch makes assumptions about the current working directory and I'm not very comfortable manipulating `sys.path`.

**Comment by kloczek** [comment]
> The project needs to be installed for this to work (even a "develop" install will do). Have you done that? 

I'm just using provided docs/Makefile. The same you can do by  executing `build-sphinx` command in project root directory like `sphinx-build -n -T -b man docs build/sphinx/man`.

> That patch makes assumptions about the current working directory and I'm not very comfortable manipulating sys.path.

No it doesn't make any assumption. `sphinx-build` command automattically during documentation rendering process changes current directory to location where it is copy.py file, so to provide path to the module `sys.path` needs to be altered or such path needs to be provided in $PYTHONPATH.
You can find such `sys.path`  injection in maaany modules copy.py files.

That path needs to be provided because for example not always foo/ module source code is in root directory of the project. Sometimes it is in src/foo/ subdirectory.
With that patch it possible to generate documentation without have installed it and **always** use project tree code (and not installed version of the module which may be different).

**Comment by kloczek** [comment]
> If you don't install the project, how will the documentation dependencies get installed? The `tox -e docs` command is a really neat way to do everything in one shot, and you should either use that or perform the same steps manually. I will remove the `Makefile` to reduce the confusion.
Maybe it is not obviouse but that tox target is installong module in venv and than used inside that venv `sphinx-build` generetes documentation out of installed inside `typeguard` module.
That patch allow build documentation like in any other project outsize pytjon modules domenn .. from spirce tree withoit uisng tox which using takes tiime to assemble that venv.
Those three lines adeed provides way simpler documentation procedure .. KISS principle.

Currently you can generate documemtation as well without tox if module code is available over `sys.path` (if `typeguard` is already installed in system image or user home path) however in such case executing `sphinx-buil` command may generate documetation for completly different version than what it is in source tree.

**Comment by agronholm** [comment]
Yeah, which is why one should use a virtual environment to do it.

**Comment by agronholm** [comment]
I know how setuptools_scm works. But none of what you just wrote helps in getting the correct version in the documentation.

**Comment by kloczek** [comment]
If you will generate .whl after that you have in source tree src/typeguard.dist-info direcory with all module metadata.
Addind that path to `sys.path` allows any code from conf.py allows read that direcory metadata.

**Comment by agronholm** [comment]
Okay, so why don't you do that and set `PYTHONPATH` accordingly?

**Comment by kloczek** [comment]
Because sometimes module is in root direcory and sometimes in src/ (or even in yet another location), and altering copy.py in each package allows me to have fixed commands sequence (rpm macro) no matter where module code in source tree is.
Submitting such copy.py and acceping such patch by maintainer allows me to remove that patch (which fixes/adds copf.py modyfication) from my build procedure.
How often such modyfication is necessary to add?
```console
[tkloczko@devel-g2v SPECS]$ grep Patch.*%{name}-add_module_path_in_copy.py.patch python-* | wc -l; grep %sphinx_build_man python-* | wc -l; ls -1 python-*|wc -l
67
501
924
````
So ~5/9  of all my modules have sphinx based documentation and amongs those only a little more than 10% needs such patch adding module path in copy,py.

**Comment by agronholm** [comment]
As of 3.0.0rc1, all Sphinx warnings are gone. I've changed the settings to fail the docs build if there are any warnings.

**Comment by kloczek** [comment]
Thank you 👍 
If may I ask next time make new releae because githib sends notyfications only about those git tags which are made as releases.
Currently there are no git repo releases at all https://github.com/agronholm/typeguard/releases
I'm askimg about tthat becaue my automation catches releases notyfications automatically trying to rebuild my rpm package with new version.
(I would not recommend to mark as releases rc/a/b tags)

Is it anyting more still on outstabding ist before release 3..0.0? 🤔

**Comment by agronholm** [comment]
There have been tons of changes in the 3.0 series which have received very little practical testing. I'll let it stew for a while and if no significant issues crop up, I will make a final release.

As far as GitHub releases go, I'm looking for proper automation that would make a release when pushing a tag. Preferably it would also extract the changes from the changelog to be part of the release. Do you happen to know a GH Action that will do that?

**Comment by agronholm** [comment]
Also, can you not set up your automation to look for new tags?

**Comment by kloczek** [comment]
> Also, can you not set up your automation to look for new tags?

Hmm .. any particular reason? 🤔

**Comment by agronholm** [comment]
So I wouldn't have to add automation for making GitHub releases? I have no other reason to do that.

**Comment by kloczek** [comment]
I don't think that you need to be worry about my automation. 😋 
Grabbing new version from new tag and try to package that with new version is only first step of the relatively long (at the moment) pipeline of approval processes to accept that as useful resource.
In that pipeline is testing such package against builds of other package are they still building and passing they test suites. Other set of checks is about install time dependencies (is new package breaks or not somehow other packages install time dependencies.
At the end still is Human™️  which is making the decision about merge package which must pass all existing checks to be part of the packages repository 🤓 
Just checked unpacked rpm spec files which I'm using:
```console
[tkloczko@pers-jacek SPECS]$ grep "BuildRequires:.*python3dist(typeguard)" *
python-nptyping.spec:BuildRequires:     python3dist(typeguard)
python-stack-data.spec:BuildRequires:   python3dist(typeguard)
python-tenacity.spec:BuildRequires:     python3dist(typeguard)
```
In other words at the moment at least three other packages builds will be automatically verified that they are still OK with new version (In this case I'm talking about "at leas" because in this case that grep catches only direct Buildrequiresa and `typeguard` may be on other packages indirect build dependencies).
In other words .. I don't thing that you have now possibility to check your new version in that way 😋 
And .. if it will be anything wrong cached by that automation which (after maual review) may point on some `typeguard` code issues definitelly will ring the bell here😃 
In the futur I may offer public access to such automation to be able use that as part of other projects CIs however it is still looooong way to open that for such use cases 😄
